src/ImageSharp/Formats/Png/PngEncoderCore.cs

```diff
@@ -1192,4 +1219,196 @@ private void SwapScanlineBuffers()
+    }
+
+    /// <summary>
+    /// Creates the quantized frame.
+    /// </summary>
+    /// <typeparam name="TPixel">The type of the pixel.</typeparam>
+    /// <param name="encoder">The png encoder.</param>
+    /// <param name="colorType">The color type.</param>
+    /// <param name="bitDepth">The bits per component.</param>
+    /// <param name="image">The image.</param>
+    private static IndexedImageFrame<TPixel> CreateQuantizedFrame<TPixel>(
+        IQuantizingEncoderOptions encoder,
+        PngColorType colorType,
+        byte bitDepth,
+        Image<TPixel> image)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        if (colorType != PngColorType.Palette)
+        {
+            return null;
+        }
+
+        // Use the metadata to determine what quantization depth to use if no quantizer has been set.
+        IQuantizer quantizer = encoder.Quantizer
+            ?? new WuQuantizer(new QuantizerOptions { MaxColors = ColorNumerics.GetColorCountForBitDepth(bitDepth) });
+
+        // Create quantized frame returning the palette and set the bit depth.
+        using IQuantizer<TPixel> frameQuantizer = quantizer.CreatePixelSpecificQuantizer<TPixel>(image.GetConfiguration());
+
+        frameQuantizer.BuildPalette(encoder.PixelSamplingStrategy, image);
+        return frameQuantizer.QuantizeFrame(image.Frames.RootFrame, image.Bounds());
+    }
+
+    /// <summary>
+    /// Calculates the bit depth value.
+    /// </summary>
+    /// <typeparam name="TPixel">The type of the pixel.</typeparam>
+    /// <param name="colorType">The color type.</param>
+    /// <param name="bitDepth">The bits per component.</param>
+    /// <param name="quantizedFrame">The quantized frame.</param>
+    /// <exception cref="NotSupportedException">Bit depth is not supported or not valid.</exception>
+    private static byte CalculateBitDepth<TPixel>(
+        PngColorType colorType,
+        byte bitDepth,
+        IndexedImageFrame<TPixel> quantizedFrame)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        if (colorType == PngColorType.Palette)
+        {
+            byte quantizedBits = (byte)Numerics.Clamp(ColorNumerics.GetBitsNeededForColorDepth(quantizedFrame.Palette.Length), 1, 8);
+            byte bits = Math.Max(bitDepth, quantizedBits);
+
+            // Png only supports in four pixel depths: 1, 2, 4, and 8 bits when using the PLTE chunk
+            // We check again for the bit depth as the bit depth of the color palette from a given quantizer might not
+            // be within the acceptable range.
+            if (bits == 3)
+            {
+                bits = 4;
+            }
+            else if (bits is >= 5 and <= 7)
+            {
+                bits = 8;
+            }
+
+            bitDepth = bits;
+        }
+
+        if (Array.IndexOf(PngConstants.ColorTypes[colorType], bitDepth) == -1)
```

Comment:
```suggestion
        if (Array.IndexOf(PngConstants.ColorTypes[colorType], bitDepth) < 0)
```
It's tiny little bit faster than `== -1` as the cpu can fuse the compare and jump into one instruction actually executed (the `test + jmp`-combo).